lizrd/core/llm.py

```diff
@@ -128,65 +130,96 @@ def LowRank(dinput, doutput, dlowrank):
+
+
 @ash.check("... d -> ... d")
-class Attention(nn.Module):
-    def __init__(self, dmodel, heads, dhead=None):
+class Attention(LoggingLayer):
+    def __init__(self, dmodel, heads, causal, dhead=None, flash=False):
         super(Attention, self).__init__()
         if dhead is None:
             assert dmodel % heads == 0
             dhead = dmodel // heads
 
         self.heads = heads
         self.dhead = dhead
-        self.dmodel = dmodel
+        self.causal = causal
+        self.flash = flash
 
-        key_query_value_gen = lambda: misc.EinMix(
-            "... dmodel -> ... heads dhead",
-            weight_shape="dmodel heads dhead",
-            bias_shape="heads dhead",
-            dmodel=dmodel,
-            heads=heads,
-            dhead=dhead,
-        )
+        self.input_projection = nn.Linear(dmodel, 3 * heads * dhead)
+        self.output_projection = nn.Linear(heads * dhead, dmodel)
 
-        self.Q = key_query_value_gen()
-        self.K = key_query_value_gen()
-        self.V = key_query_value_gen()
-
-        combine_gen = lambda: misc.EinMix(
-            "... heads dhead -> ... dmodel",
-            weight_shape="heads dhead dmodel",
-            bias_shape="dmodel",
-            dmodel=dmodel,
-            heads=heads,
-            dhead=dhead,
+    def forward(self, x):
+        projected = self.input_projection(x)
+
+        batch, seq_len = x.shape[:-1]
+        projected = projected.view(
+            batch, seq_len, self.heads, 3 * self.dhead
+        ).transpose(1, 2)
+        q, k, v = torch.chunk(projected, chunks=3, dim=-1)
+
+        attention_output = attention_mechanism(
+            query=q,
+            key=k,
+            value=v,
+            dhead=self.dhead,
+            causal=self.causal,
+            flash=self.flash,
         )
-        self.D = combine_gen()
 
-    def forward(self, x):
-        q = self.Q(x)
-        k = self.K(x)
-        v = self.V(x)
+        output = self.output_projection(attention_output.transpose(1, 2).flatten(-2))
 
-        a = torch.einsum("... l h d, ... L h d -> ... h l L", q, k)
-        a = a * (1 / self.dhead**0.5)
-        a = torch.softmax(a, dim=-1)
-        prefinal = torch.einsum("... h l L, ... L h d -> ... l h d", a, v)
-        output = self.D(prefinal)
         return output
 
 
 @ash.check("... d -> ... d")
-class CausalAttention(nn.Module):
-    def __init__(self, dmodel, heads, dhead=None):
-        super(CausalAttention, self).__init__()
+class LegacyAttention(nn.Module):
+    def __init__(self, dmodel, heads, causal, flash, dhead=None):
+        super(LegacyAttention, self).__init__()
         if dhead is None:
             assert dmodel % heads == 0
             dhead = dmodel // heads
 
         self.heads = heads
         self.dhead = dhead
         self.dmodel = dmodel
+        self.causal = causal
+        self.flash = flash
 
         key_query_value_gen = lambda: misc.EinMix(
             "... dmodel -> ... heads dhead",
```

Comment:
```suggestion
        key_query_value_gen = lambda x: x.view(x.size(0), x.size(1), self.heads, self.dhead)
```

Reply:
This code is just copied from old version of Attention. I see no point in changing it - it's just legacy version that will be removed in one week. Let's keep it as is

Reply:
ok!